### PR TITLE
keycloak_userprofile: new module

### LINKED ISSUE #8573: Community.General should support Keycloak 24 new Realm setting 'unmanagedAttributePolicy'
### Summary

Becase right now it is NOT possible to change Keycloak setting 'unmanagedAttributePolicy',
which appeared in Keycloak 24 Realm configuration.
This ('unmanagedAttributePolicy') Is an important Realm configuration property, but it can't be changed by Community.General now.

See Keycloak documentation : 
https://www.keycloak.org/docs/latest/upgrading/index.html#user-profile-changes
https://www.keycloak.org/docs/latest/server_admin/#_understanding-managed-and-unmanaged-attributes
https://www.keycloak.org/docs/latest/server_admin/#_user-profile-json-configuration   

Also here is Terraform Keycloak fixed the same issue of new 'unmanagedAttributePolicy' : 
https://github.com/mrparkers/terraform-provider-keycloak/issues/946

### Issue Type

Bug Report

### Component Name

Keycloak

### Ansible Version

the latest

### Community.general Version

the latest

### Configuration

not relevant

### OS / Environment

not relevant

### Steps to Reproduce

there is simply no way to change Keycloak  24 (new) 'unmanagedAttributePolicy' because it is NOT exposed

### Expected Results

able to change Keycloak  24 (new) 'unmanagedAttributePolicy' 

### Actual Results

not relevant

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by yeoldegrove**
##### SUMMARY

This adds a new keycloak module to manage user profiles.

See Keycloak documentation :
https://www.keycloak.org/docs/latest/upgrading/index.html#user-profile-changes
https://www.keycloak.org/docs/latest/server_admin/#_understanding-managed-and-unmanaged-attributes
https://www.keycloak.org/docs/latest/server_admin/#_user-profile-json-configuration

Fixes #8573 and #8579

##### ISSUE TYPE

- New Module/Plugin Pull Request

##### COMPONENT NAME

keycloak_userprofile

##### ADDITIONAL INFORMATION

This is heavily based on `keycloak_realm_key` and `keycloak_user_federation`.

### CODE REVIEW

**Comment by yeoldegrove** [comment]
Unit tests still do not work (hints welcome).
Some initial feedback would be appreciated, though.

File: tests/unit/plugins/modules/test_keycloak_userprofile.py [hunk 1]
```diff
@@ -0,0 +1,998 @@
...
+
+        # Run the module
+
+        with mock_good_connection():
+            with patch_keycloak_api(get_components=return_value_get_components_get, get_component=return_value_component_get,
+                                    update_component=return_value_component_update) as (mock_get_components, mock_get_component, mock_create_component, mock_update_component, mock_delete_component):
+                with self.assertRaises(AnsibleExitJson) as exec_info:
+                    self.module.main()
+
+        self.assertEqual(len(mock_get_components.mock_calls), 2)
```
Comment (felixfontein) [inline tests/unit/plugins/modules/test_keycloak_userprofile.py:913]: `ks.get_components` is only ever called once in the module (line 510, outside of a loop). Why do you expect it to be called twice?

Comment (yeoldegrove) [reply tests/unit/plugins/modules/test_keycloak_userprofile.py]: @felixfontein Thanks for the hint! After looking at it again I could make sense of the unit test ;)
Also I fixed all of the remaining sanity checks and reworked the integration tests.

I think this is good for a review.

**Comment by yeoldegrove** [comment]
@felixfontein Anything else needed so this can be merged?

File: tests/sanity/ignore-2.18.txt [hunk 2]
```diff
@@ -2,6 +2,7 @@ plugins/modules/consul_session.py validate-modules:parameter-state-invalid-choic
 plugins/modules/homectl.py import-3.11  # Uses deprecated stdlib library 'crypt'
 plugins/modules/homectl.py import-3.12  # Uses deprecated stdlib library 'crypt'
 plugins/modules/iptables_state.py validate-modules:undocumented-parameter             # params _back and _timeout used by action plugin
+plugins/modules/keycloak_userprofile.py validate-modules:parameter-invalid
```
Comment (felixfontein) [inline tests/sanity/ignore-2.18.txt:5]: Why is this needed? Adding new entries here should be done with care, and this one looks like it is hiding a bug in the module.

File: plugins/modules/keycloak_userprofile.py [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: keycloak_userprofile
+
+short_description: Allows managing Keycloak User Profiles.
```
Comment (felixfontein) [inline plugins/modules/keycloak_userprofile.py:15]: This is the only description which shouldn't end with a period :)
```suggestion
short_description: Allows managing Keycloak User Profiles
```

File: plugins/modules/keycloak_userprofile.py [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+                            type='list',
+                            elements='dict',
+                            required=False,
+                            options={
+                                'name': dict(type='str', required=True),
+                                'displayName': dict(type='str', required=True),
+                                'validations': dict(
+                                    type='dict',
+                                    # TODO: template this too but without passing null on unrequired empty key/value pairs
+                                    # options={
```
Comment (felixfontein) [inline plugins/modules/keycloak_userprofile.py:384]: Basically if `options` isn't added now, it can never be added in the future.

What exactly is the problem here? That Ansible automatically adds `None` values for all fields the user did not provide? If that's the problem, you have to filter out these values yourself.

File: tests/sanity/ignore-2.18.txt (hunk 2 above)
Comment (yeoldegrove) [reply tests/sanity/ignore-2.18.txt:5]: The concrete problem is the suboption `kc.user.profile.config` which is "so not python". I could of course abstract/rename this to be more pythonesque, but this is how it is named in keycloak's json representation.
Do you think it is valid to leave it or should it be changed?

Comment (felixfontein) [reply tests/sanity/ignore-2.18.txt:5]: How about naming that option `kc_user_profile_config`, with an alias `kc.user.profile.config`? That way users can use both variants. The code will need to handle `kc_user_profile_config` though (and move its value to `kc.user.profile.config` when using `module.params['config']`), something like:
```
    config = module.params['config']
    if config:
        config = dict(config)  #  create a copy
        config['kc.user.profile.config'] = config.pop(`kc_user_profile_config`)
```
You're using `module.params['config']` (or its eqiuvalent `module.params.get('config')`) further down the code as well though, that needs to be adjusted to use this copy.

**Comment by felixfontein** [comment]
Ping @yeoldegrove

File: tests/sanity/ignore-2.18.txt (hunk 2 above)
Comment (yeoldegrove) [reply tests/sanity/ignore-2.18.txt:5]: Implemented your suggestion to name it `kc_user_profile_config` and now have aliases for a lot of (sub-)options.
Also added code that handles recursive camelCasel lists and dicts.

File: plugins/modules/keycloak_userprofile.py (hunk 3 above)
Comment (yeoldegrove) [reply plugins/modules/keycloak_userprofile.py]: :+1:

**Comment by yeoldegrove** [comment]
@felixfontein Finally had time to work on this again. I think I tackled everything you mentioned.

File: tests/sanity/ignore-2.18.txt (hunk 2 above)
Comment (russoz) [reply tests/sanity/ignore-2.18.txt:5]: I think it would be better not to have the name with the dots exposed to the Ansible user. It is not just the non- pythonesque aspect of it, but it would also create problems if that comes up in Jinja templating. Not to mention the need to create these exceptions in the sanity checking.
The problems outweigh the benefits, IMHO.

File: tests/unit/plugins/modules/test_keycloak_userprofile.py [hunk 5]
```diff
@@ -0,0 +1,1001 @@
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2021, Ansible Project
```
Comment (russoz) [inline tests/unit/plugins/modules/test_keycloak_userprofile.py:3]: Copy paste quirks ;-)
```suggestion
# Copyright (c) 2024, Ansible Project
```

**Review by russoz — COMMENTED** [review]
Hi @yeoldegrove , thanks for your contribution. Overall it does look good, there's just a couple of comments below, mostly minor issues.

File: tests/sanity/ignore-2.18.txt (hunk 2 above)
Comment (yeoldegrove) [reply tests/sanity/ignore-2.18.txt]: Alright, removed the "doted" parameter option completely.

File: plugins/modules/keycloak_userprofile.py [hunk 6]
```diff
@@ -0,0 +1,668 @@
...
+    parent_id:
+        description:
+            - The parent_id of the realm userprofile. In practice, the ID (name) of the realm.
+        aliases:
+            - parentId
+        type: str
+        required: true
+    provider_id:
+        description:
+            - The name of the "provider ID" for the userprofile.
```
[the remark this answers is not in the document]
Comment (yeoldegrove) [reply plugins/modules/keycloak_userprofile.py]: Adapted things to have more consistent naming.

File: plugins/modules/keycloak_userprofile.py [hunk 7]
```diff
@@ -0,0 +1,668 @@
...
+        aliases:
+            - providerId
+        choices: ['declarative-user-profile']
+        default: 'declarative-user-profile'
+        type: str
+    config:
+        description:
+            - The configuration of the User Profile Provider.
```
[the remark this answers is not in the document]
Comment (yeoldegrove) [reply plugins/modules/keycloak_userprofile.py:74]: Adapted things to have more consistent naming.

File: plugins/modules/keycloak_userprofile.py [hunk 8]
```diff
@@ -0,0 +1,668 @@
...
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.six.moves.urllib.parse import urlencode
+from copy import deepcopy
+import json
+
+
+def remove_null_values(data):
+    if isinstance(data, dict):
+        # Recursively remove null values from dictionaries
+        return {k: remove_null_values(v) for k, v in data.items() if v is not None}
```
[the remark this answers is not in the document]
Comment (yeoldegrove) [reply plugins/modules/keycloak_userprofile.py:418]: :tada:

File: plugins/modules/keycloak_userprofile.py [hunk 9]
```diff
@@ -0,0 +1,668 @@
...
+        module.fail_json(msg=str(e))
+
+    kc = KeycloakAPI(module, connection_header)
+
+    params_to_ignore = list(keycloak_argument_spec().keys()) + ["state"]
+
+    # Filter and map the parameters names that apply to the role
+    component_params = [x for x in module.params
+                        if x not in params_to_ignore and
+                        module.params.get(x) is not None]
```
[the remark this answers is not in the document]
Comment (yeoldegrove) [reply plugins/modules/keycloak_userprofile.py]: :+1:

File: tests/sanity/ignore-2.13.txt [hunk 10]
```diff
@@ -5,6 +5,7 @@ plugins/lookup/etcd.py validate-modules:invalid-documentation
 plugins/lookup/etcd3.py validate-modules:invalid-documentation
 plugins/modules/consul_session.py validate-modules:parameter-state-invalid-choice
 plugins/modules/iptables_state.py validate-modules:undocumented-parameter             # params _back and _timeout used by action plugin
+plugins/modules/keycloak_userprofile.py validate-modules:parameter-invalid
```
[the remark this answers is not in the document]
Comment (yeoldegrove) [reply tests/sanity/ignore-2.13.txt]: Removed the ignores completely. This was initially about the doted `kc.user.profile.config` parameter.

File: tests/integration/targets/keycloak_userprofile/readme.adoc
[the remark this answers is not in the document]
Comment (yeoldegrove) [reply tests/integration/targets/keycloak_userprofile/readme.adoc:1]: Set up the start of the container, too now.

**Comment by yeoldegrove** [comment]
@russoz Thanks for the review! Reworked everything you mentioned and also re-shuffled a few things around with `partent_id` and `realm` and fixed some warnings and errors in the unit and integration tests.

File: plugins/modules/keycloak_userprofile.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (felixfontein) [reply plugins/modules/keycloak_userprofile.py:418]: Yep, we recently dropped Python 2.6 support :) (And also Python 2.7 support on the controller side - but for modules it's still supported.)

I'm looking forward to dropping Python 2.7 support for modules as well, IIRC that will happen somewhen next year (once the collection drops support for ansible-core 2.16).

**Comment by felixfontein** [comment]
Looks ok to me. If nobody objects, I'll merge this for 9.4.0 on this weekend.

File: plugins/modules/keycloak_userprofile.py [hunk 11]
```diff
@@ -0,0 +1,727 @@
...
+                          - unmanagedAttributePolicy
+                        type: str
+                        choices:
+                            - ENABLED
+                            - ADMIN_EDIT
+                            - ADMIN_VIEW
+
+notes:
+    - Currently only the provider_id of declarative-user-profile is tested.
+    - Currently only a single declarative-user-profile entry can exist. This entry can have multiple attributes, though.
```
Comment (russoz) [inline plugins/modules/keycloak_userprofile.py:270]: These two lines should probably used semantic markup as well, but, the first one is quite redundant with the description in the option itself, on line 56. Maybe just move the second line to the description up there, put the V() around the value, and remove the first?

Comment (yeoldegrove) [reply plugins/modules/keycloak_userprofile.py]: @russoz rephrased the notes section and it uses semantic markup now as well.

File: changelogs/fragments/8651-keycloak-userprofile.yml
Comment (felixfontein) [inline changelogs/fragments/8651-keycloak-userprofile.yml:1]: Please remove this file, new modules should not come with changelog fragments. They are automatically announced via their `version_added` version.

File: plugins/modules/keycloak_userprofile.py [hunk 12]
```diff
@@ -0,0 +1,727 @@
...
+            - The parent id of the realm key. In practice the ID (name) of the realm.
+        aliases:
+            - parentId
+            - realm
+        type: str
+        required: true
+
+    provider_id:
+        description:
+            - The name of the provider id for the key (supported value is V(declarative-user-profile)).
```
Comment (felixfontein) [inline plugins/modules/keycloak_userprofile.py:56]:
```suggestion
            - The name of the provider ID for the key (supported value is V(declarative-user-profile)).
```

File: plugins/modules/keycloak_userprofile.py [hunk 13]
```diff
@@ -0,0 +1,727 @@
...
+                            annotations:
+                                description:
+                                    - The annotations included in the group.
+                                type: dict
+                                required: false
+
+                    unmanaged_attribute_policy:
+                        description:
+                            - Policy for unmanaged attributes.
+                            - Supported values are V(ENABLED), V(ADMIN_EDIT) and V(ADMIN_VIEW).
```
Comment (felixfontein) [inline plugins/modules/keycloak_userprofile.py:259]: These are already listed under `choices`, so there's no need to list them again.
```suggestion
```

File: plugins/modules/keycloak_userprofile.py [hunk 14]
```diff
@@ -0,0 +1,727 @@
...
+                          - unmanagedAttributePolicy
+                        type: str
+                        choices:
+                            - ENABLED
+                            - ADMIN_EDIT
+                            - ADMIN_VIEW
+
+notes:
+    - Currently, only a single V(declarative-user-profile) entry is supported for O(provider_id) (design of the Keyckoak API).
+    - However, there can be multiple O(config.kc_user_profile_config[].attributes[]) entries.
```
Comment (felixfontein) [inline plugins/modules/keycloak_userprofile.py:270]: I think this should be part of the previous note.
```suggestion
      However, there can be multiple O(config.kc_user_profile_config[].attributes[]) entries.
```

File: plugins/modules/keycloak_userprofile.py [hunk 15]
```diff
@@ -0,0 +1,727 @@
...
+
+- name: Delete a Keycloak User Profile Provider
+  keycloak_userprofile:
+    state: absent
+    parent_id: master
+
+# Unmanaged attributes are user attributes not explicitly defined in the User Profile configuration. By default, unmanaged attributes are `Disabled`
+# and are not available from any context such as registration, account, and the administration console. By setting `Enabled`, unmanaged attributes are fully
+# recognized by the server and accessible through all contexts, useful if you are starting migrating an existing realm to the declarative User Profile
+# and you don't have yet all user attributes defined in the User Profile configuration.
```
Comment (felixfontein) [inline plugins/modules/keycloak_userprofile.py:365]: Better limit the line length in EXAMPLES, otherwise the user has to scroll horizontally to be able to read this. Also better use regular quotes than backticks.
```suggestion
# Unmanaged attributes are user attributes not explicitly defined in the User Profile
# configuration. By default, unmanaged attributes are "Disabled" and are not
# available from any context such as registration, account, and the
# administration console. By setting "Enabled", unmanaged attributes are fully
# recognized by the server and accessible through all contexts, useful if you are
# starting migrating an existing realm to the declarative User Profile
# and you don't have yet all user attributes defined in the User Profile configuration.
```

File: plugins/modules/keycloak_userprofile.py [hunk 16]
```diff
@@ -0,0 +1,727 @@
...
+- name: Enable Unmanaged Attributes
+  community.general.keycloak_userprofile:
+    state: present
+    parent_id: master
+    config:
+      kc_user_profile_config:
+        - unmanagedAttributePolicy: ENABLED
+
+# By setting `Only administrators can write`, unmanaged attributes can be managed only through the administration console and API, useful if you have already
+# defined any custom attribute that can be managed by users but you are unsure about adding other attributes that should only be managed by administrators.
```
Comment (felixfontein) [inline plugins/modules/keycloak_userprofile.py:375]:
```suggestion
# By setting "Only administrators can write", unmanaged attributes can be managed
# only through the administration console and API, useful if you have already
# defined any custom attribute that can be managed by users but you are unsure
# about adding other attributes that should only be managed by administrators.
```

File: plugins/modules/keycloak_userprofile.py [hunk 17]
```diff
@@ -0,0 +1,727 @@
...
+- name: Enable ADMIN_EDIT on Unmanaged Attributes
+  community.general.keycloak_userprofile:
+    state: present
+    parent_id: master
+    config:
+      kc_user_profile_config:
+        - unmanagedAttributePolicy: ADMIN_EDIT
+
+# By setting `Only administrators can view`, unmanaged attributes are read-only and only available through the administration console and API.
```
Comment (felixfontein) [inline plugins/modules/keycloak_userprofile.py:384]:
```suggestion
# By setting "Only administrators can view", unmanaged attributes are read-only
# and only available through the administration console and API.
```

File: changelogs/fragments/8651-keycloak-userprofile.yml
Comment (yeoldegrove) [reply changelogs/fragments/8651-keycloak-userprofile.yml:1]: :zap:

File: plugins/modules/keycloak_userprofile.py (hunk 12 above)
Comment (yeoldegrove) [reply plugins/modules/keycloak_userprofile.py:56]: :+1:

File: plugins/modules/keycloak_userprofile.py (hunk 13 above)
Comment (yeoldegrove) [reply plugins/modules/keycloak_userprofile.py:259]: :+1:

File: plugins/modules/keycloak_userprofile.py (hunk 14 above)
Comment (yeoldegrove) [reply plugins/modules/keycloak_userprofile.py:270]: :+1:

File: plugins/modules/keycloak_userprofile.py (hunk 15 above)
Comment (yeoldegrove) [reply plugins/modules/keycloak_userprofile.py:365]: :+1:

File: plugins/modules/keycloak_userprofile.py (hunk 17 above)
Comment (yeoldegrove) [reply plugins/modules/keycloak_userprofile.py:384]: :+1:

**Comment by yeoldegrove** [comment]
@felixfontein Fixed your last suggestions.

**Comment by felixfontein** [comment]
FYI @yeoldegrove you might be interested in https://forum.ansible.com/t/keycloak-modules-in-community-general/41746

**Comment by russoz** [comment]
Apparently, upon reading this:

> Fixes #8573 and #8579 

GitHub associates only the first issue, the second one gets only a mention.
